tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla

```diff
@@ -0,0 +1,446 @@
+    /\ diskPos = 0
+    /\ file_pointer = 0
+    /\ file_content = EmptyArray
+
+FlushBuffer ==
+    /\ dirty
+    /\ LET len == Min(length - lo, BuffSz) IN
+        /\ IF len > 0
+           THEN LET diskPosA == lo IN \* super.seek(this.lo)
+            /\ file_content' = WriteToFile(file_content, diskPosA, ArraySlice(buff, 0, len))
+            /\ file_pointer' = diskPosA + len
+            /\ diskPos' = lo + len
+           ELSE
+            UNCHANGED <<diskPos, file_pointer, file_content>>
+        /\ dirty' = FALSE
+    /\ UNCHANGED <<length, curr, lo, buff>>
+
+\* Helper for Seek (not a full action):
+\*  - reads lo'
+\*  - constrains diskPos', file_pointer', and buff'
+FillBuffer ==
+    LET diskPosA == lo' IN
+    /\ buff' = MkArray(BuffSz, [i \in 0..BuffSz |->
+            LET fileOffset == diskPosA + i IN
+            IF fileOffset < ArrayLen(file_content)
+            THEN ArrayGet(file_content, fileOffset)
+            ELSE ArbitrarySymbol])
+    /\ file_pointer' = Min(diskPosA + BuffSz, ArrayLen(file_content))
+    /\ diskPos' = Min(diskPosA + BuffSz, ArrayLen(file_content))
+
+Seek(pos) ==
+    /\ curr' = pos
+    /\ IF pos < lo \/ pos >= (lo + BuffSz) THEN
+        /\ ~dirty \* call to FlushBuffer
+        /\ lo' = (pos \div BuffSz) * BuffSz
+        /\ FillBuffer
+       ELSE
+        UNCHANGED <<lo, diskPos, file_pointer, buff>>
+    /\ UNCHANGED <<dirty, length, file_content>>
+
+SetLength(newLength) ==
+    /\ file_content' = TruncateOrExtendFile(file_content, newLength)
+    /\ IF ArrayLen(file_content) > newLength /\ file_pointer > newLength
+       THEN file_pointer' = newLength
+       ELSE file_pointer' \in Offset
+    /\ length' = newLength
+    /\ diskPos' = file_pointer'
+    /\ IF curr > newLength
+       THEN curr' = newLength
+       ELSE UNCHANGED curr
+    \* In reality the buffer doesn't change---but some of its bytes might no
+    \* longer be relevant and have to be marked as arbitrary.
+    /\ buff' = MkArray(BuffSz, [i \in 0..(BuffSz-1) |->
+            IF lo + i < newLength
+            THEN ArrayGet(buff, i)
+            ELSE ArbitrarySymbol])
+    /\ UNCHANGED <<dirty, lo>>
+
+Read1(byte) ==
+    /\ Inv2
+    /\ curr < length
+    /\ byte = ArrayGet(buff, curr - lo)
+    /\ curr' = curr + 1
+    /\ UNCHANGED <<lo, diskPos, buff, file_pointer, dirty, file_content, length>>
+
+Write1(byte) ==
+    /\ curr + 1 <= MaxOffset \* bound model checking
+    /\ Inv2
+    /\ buff' = ArraySet(buff, curr - lo, byte)
+    /\ curr' = curr + 1
+    /\ dirty' = TRUE
+    /\ length' = Max(length, curr')
+    /\ UNCHANGED <<lo, diskPos, file_pointer, file_content>>
+
+Read(data) ==
+    LET numReadableWithoutSeeking == Min(lo + BuffSz, length) - curr IN
+    /\ Inv2
+    /\ numReadableWithoutSeeking >= 0
+    /\ LET numToRead == Min(ArrayLen(data), numReadableWithoutSeeking) IN
+       LET buffOff == curr - lo IN
+        /\ data = ArraySlice(buff, buffOff, buffOff + numToRead)
```

Comment:
Neither `Read` nor `Read1` are enabled if the buffer contains `ArbitrarySymbol`, i.e., the following invariant holds:

```tla
DebugReadsWithArbitrarySymbol ==
    \E i \in 1..Len(buff.elems): buff.elems[i] = ArbitrarySymbol =>
        /\ \A data \in UNION{ Array(Symbols, len): len \in 1..MaxOffset }:
                ~ENABLED Read(data)
        /\ \A symbol \in Symbols:
                ~ENABLED Read1(symbol) 
```

Why this restriction, given that `BufferedRandomAccessFile.java` does not impose such a restriction?

Reply:
You are right; I believe that restriction can and should be relaxed.

Reply:
@lemmy There is an associativity problem in your definition; I think it should be
```tla
DebugReadsWithArbitrarySymbol ==
    (\E i \in 1..Len(buff.elems): buff.elems[i] = ArbitrarySymbol) => \* NOTE parens around \E expression
        /\ ...
```

Using that modified definition shows that `Read` actually is enabled when the read does not cover `ArbitrarySymbol`, even if there are `ArbitrarySymbol`s in the buffer.

Even so, I do think it makes sense to allow reads of `ArbitrarySymbol` in the next-relation.